### DISCUSSION TITLE
The rappid.tgz file is not added  to the node_modules folder during `npm install`

### DISCUSSION THREAD

**Discussion by martijn-van-den-bergh** [clientIO/joint#2192] [Q&A] [answered]
### Introduction

I am trying to install the package from the developer portal. I have version 3.6.3. But when I run npm install and I check the node_modules folder i do not see a `@clientio/rappid` folder.

The tgz file is in the root folder of an Angular project and my package.json looks as follows:

```
{
    "name": "name",
    "version": "9.9.0",
    "scripts": {
...
    },
    "author": "Author",
    "license": "ISC",
    "private": true,
    "dependencies": {
        "@angular/animations": "~11.2.5",
        "@angular/cdk": "^11.2.6",
        "@angular/common": "~11.2.5",
        "@angular/compiler": "~11.2.5",
        "@angular/core": "~11.2.5",
        "@angular/forms": "~11.2.5",
        "@angular/material": "^11.2.6",
        "@angular/material-moment-adapter": "^12.0.5",
        "@angular/platform-browser": "~11.2.5",
        "@angular/platform-browser-dynamic": "~11.2.5",
        "@angular/router": "~11.2.5",
        "@clientio/rappid": "file:rappid.tgz",
        "dagre": "0.8.4"
    },
    "devDependencies": {
        "@angular-devkit/build-angular": "~0.1102.3",
        "@angular/cli": "~11.2.4",
        "@angular/compiler-cli": "~11.2.5",
        "@angular/localize": "^11.2.7",
        "@types/jasmine": "~3.6.0",
        "@types/node": "^12.11.1",
        "codelyzer": "^6.0.0",
        "jasmine-core": "~3.6.0",
        "jasmine-spec-reporter": "~5.0.0",
        "karma": "~6.1.0",
        "karma-chrome-launcher": "~3.1.0",
        "karma-coverage": "~2.0.3",
        "karma-jasmine": "~4.0.0",
        "karma-jasmine-html-reporter": "^1.5.0",
        "karma-junit-reporter": "^2.0.1",
        "protractor": "~7.0.0",
        "ts-node": "~8.3.0",
        "tslint": "~6.1.0",
        "typescript": "~4.1.5",
        "webpack-bundle-analyzer": "^4.4.1",
        "@types/backbone": "latest",
        "@types/jquery": "latest"
    }
}
```

### Steps to reproduce

I do the following:
* Delete node_modules
* Delete package-lock
* Replace the old rappig.tgz file (free trial file) with the new one from the developer portal.
* Run `npm install`

### Restrictions & Constraints

I can just copy the bundles to the @clientio/rappid folder but our CI/CD pipeline failes because it needs to install it from the tgz file. It worked fine with the free trial version but now it doesn't anymore with the payed version.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by kumilingus** [comment]
Could you please try to run `npm cache clean --force`  before `npm install`?

**Reply by martijn-van-den-bergh** [reply]
Oh, sorry. I should have mentioned that I tried that too. 

Also in the logging of the Azure DevOps Pipeline i get the following:

```
npm WARN tarball tarball data for @clientio/rappid@file:rappid.tgz (sha512-rQtiEOHAaDctzOwm05bz9iXMWaiqbHC3sLThHNr+0E/lRdUhz03ma8U/B5jwYJ4xidL01cMhTB4LZLHzcmurbw==) seems to be corrupted. Trying one more time.
  npm WARN tarball tarball data for @clientio/rappid@file:rappid.tgz (sha512-rQtiEOHAaDctzOwm05bz9iXMWaiqbHC3sLThHNr+0E/lRdUhz03ma8U/B5jwYJ4xidL01cMhTB4LZLHzcmurbw==) seems to be corrupted. Trying one more time.

```

I dont see this locally, but I think our pipeline has more logging enabled or something.

**Reply by kumilingus** [reply]
It seems that the old `package-lock.json` is still there. The original trial package `sha512` is saved in `package-lock.json` or cache.

**Reply by martijn-van-den-bergh** [reply]
Hmm, okay. I will have to check that. Thanks for the quick reply!

**Reply by martijn-van-den-bergh** [reply]
I only had the chance to look at this today. I tried with the new 3.7 version of rappid.tgz.
I did the follow steps:
1. delete node_modules
2. delete package-lock
3. run `npm cache clean --force`
4. run `npm install`

After that i have a fresh node_modules folder. But there is no sign of the @clientio/rappid folder.

If i try to run a dev server by doing the command `ng serve` i get the following error message:
`no such file or directory, open '.../AngularApp/node_modules/@clientio/rappid/index.ts'`

**Reply by kumilingus** [reply]
Does `npm install` report any errors?

**Reply by martijn-van-den-bergh** [reply]
I added --loglevel verbose but couldn't really find anything.

After that i tried npm audit and got the following:
```
npm ERR! code ELOCKVERIFY
npm ERR! Errors were found in your package-lock.json, run  npm install  to fix them.
npm ERR!     Missing: @clientio/rappid@file:rappid.tgz
```

This is after I deleted package-lock and node_modules and ran cache clean --force.
The rappid.tgz is in the exact same location as the free trial version and is imported like this:

```
{
...
    "dependencies": {
        "@angular/animations": "~11.2.5",
        "@angular/cdk": "^11.2.6",
        "@angular/common": "~11.2.5",
        "@angular/compiler": "~11.2.5",
        "@angular/core": "~11.2.5",
        "@angular/forms": "~11.2.5",
        "@angular/material": "^11.2.6",
        "@angular/material-moment-adapter": "^12.0.5",
        "@angular/platform-browser": "~11.2.5",
        "@angular/platform-browser-dynamic": "~11.2.5",
        "@angular/router": "~11.2.5",
        "@clientio/rappid": "file:rappid.tgz",
        "dagre": "0.8.4"
    },
    "devDependencies": {
        "@types/backbone": "latest",
        "@types/jquery": "latest"
    }
}
```

The rappid.tgz file is in the source folder of the angular project

**Reply by kumilingus** [reply]
What if you add a new dependency called `test123`? Can you locate it under the `node_modules/test123`?
```js
  "dependencies": {
    "test123": "file:rappid.tgz"
  }
```

**Reply by martijn-van-den-bergh** [reply]
No when I use your example I also don't get a `test123` folder in the `node_modules`.

If I run `npm audit` i see the following:

```
npm ERR! code ELOCKVERIFY
npm ERR! Errors were found in your package-lock.json, run  npm install  to fix them.
npm ERR!     Missing: test123@file:rappid.tgz
```

Though if I search in the package-lock i can not find anything regarding `test123` or `rappid`.

**Reply by kumilingus** [reply]
Then the `rappid.tgz` file must be missing or corrupted (I assume you have taken the package from the `rappid/build/package` folder from the `rappid_v3_7_0.zip` archive downloaded from `my.jointjs.com`). It's just a tarball - an archive file - you can try to unpack it from the command line to verify that the file is ok. It's the same kind of file as the trial package.

Have you tried to install it from the `rappid.tgz` from the trial package again?

**Reply by martijn-van-den-bergh** [reply, marked as answer]
I finally figured out what it was. For legacy reasons we still have an old version of node.js installed. I need to switch to that version sometimes with [NVM](https://github.com/nvm-sh/nvm). 

The old version is 10.15.0, and I was still using that to install the package. 

When I switched to a more current one, 14.17.0, it worked fine.

Thanks for thinking with me, it helped get me the solution!

**Comment by cocoa-1012** [comment]
@kumilingus @martijn-van-den-bergh 
Thanks for your this article.
How can I get the file `rappid.tgz`?

**Reply by jamesgeorgewilliams** [reply]
You can sign up for a trial:
https://www.jointjs.com/free-trial

If you are already a user of JointJS+, and have a JointJS+ package, the `rappid.tgz` is located in the `build/package` directory of your package.

Please create a new discussion if you have any further questions unrelated to this.

**Comment by cocoa-1012** [comment]
Thanks for your reply.
But I couldn't sign up with my gmail.

**Reply by kumilingus** [reply]
Could you use your company email?

**Reply by cocoa-1012** [reply]
Just used my personal email.
